flink-tests/src/test/java/org/apache/flink/test/checkpointing/RescaleCheckpointManuallyITCase.java

```diff
@@ -150,54 +137,33 @@ public void testCheckpointRescalingKeyedState(boolean scaleOut) throws Exception
                 numberElements + numberElements2,
-                client,
+                miniCluster,
                 checkpointPath);
     }
 
-    private static String runJobAndGetCheckpoint(
+    private String runJobAndGetCheckpoint(
             int numberKeys,
             int numberElements,
             int parallelism,
             int maxParallelism,
-            ClusterClient<?> client,
-            File checkpointDir)
+            MiniCluster miniCluster)
             throws Exception {
         try {
-            Duration timeout = Duration.ofMinutes(5);
-            Deadline deadline = Deadline.now().plus(timeout);
-
             JobGraph jobGraph =
                     createJobGraphWithKeyedState(
-                            parallelism, maxParallelism, numberKeys, numberElements, false, 100);
-            client.submitJob(jobGraph).get();
-
-            assertTrue(
-                    SubtaskIndexFlatMapper.workCompletedLatch.await(
-                            deadline.timeLeft().toMillis(), TimeUnit.MILLISECONDS));
-
-            // verify the current state
-            Set<Tuple2<Integer, Integer>> actualResult = CollectionSink.getElementsSet();
-
-            Set<Tuple2<Integer, Integer>> expectedResult = new HashSet<>();
-
-            for (int key = 0; key < numberKeys; key++) {
-                int keyGroupIndex = KeyGroupRangeAssignment.assignToKeyGroup(key, maxParallelism);
-                expectedResult.add(
-                        Tuple2.of(
-                                KeyGroupRangeAssignment.computeOperatorIndexForKeyGroup(
-                                        maxParallelism, parallelism, keyGroupIndex),
-                                numberElements * key));
-            }
-
-            assertEquals(expectedResult, actualResult);
-
-            // ensure contents of state within SubtaskIndexFlatMapper are all included
-            // in the last checkpoint (refer to FLINK-26882 for more details).
-            cluster.getMiniCluster().triggerCheckpoint(jobGraph.getJobID()).get();
-
-            client.cancel(jobGraph.getJobID()).get();
-            TestUtils.waitUntilJobCanceled(jobGraph.getJobID(), client);
-            return TestUtils.getMostRecentCompletedCheckpoint(checkpointDir).getAbsolutePath();
+                            parallelism,
+                            maxParallelism,
+                            numberKeys,
+                            numberElements,
+                            numberElements,
+                            true,
+                            100);
+            miniCluster.submitJob(jobGraph).get();
+            miniCluster.requestJobResult(jobGraph.getJobID()).get();
+            // The elements may not all be sent to sink when unaligned checkpoints enabled(refer to
+            // FLINK-26882 for more details).
+            // Don't verify current state here.
+            return getLatestCompletedCheckpointPath(jobGraph.getJobID(), miniCluster).get();
```

Comment:
I think the test logic has changed here. Previously, we would verify the results before restoring, however, we will only verify the results after restoring. Thus, we might not ensure all the data stored in the RocksDB state-backend, some data might still stay in the channel.

Reply:
BTW, the returned result of `getLatestCompletedCheckpointPath` might be empty.

Reply:
> Thus, we might not ensure all the data stored in the RocksDB state-backend, some data might still stay in the channel

Yes, so I delete the verification before restoring.

> the returned result of getLatestCompletedCheckpointPath might be empty.

If the result is empty, it would throw an exception here. And I think the probability is very small. Because:
1. `NotifyingDefiniteKeySource` wouldn't terminate until one checkpoint is completed. 
1. `getLatestCompletedCheckpointPath` uses `CheckpointStatsSnapshot.getHistory()` to get checkpoint path, instead of listing files under checkpointDir, as long as checkpoint is put into `CheckpointStatsHistory` after completion, `getLatestCompletedCheckpointPath` can get it.

Reply:
Can we keep the verifying logic just as before? I think that can be said as a "refactor".

Reply:
> Can we keep the verifying logic just as before? 

Because some data might still stay in the channel when unaligned checkpoint is enabled, it's hard to verify the result before restoring, and I also think the verification is not necessary, It is enough to just guarantee that the final result is correct. Thus, I changed the description.

Reply:
Thanks for the clarification.
I tried to run the test with artificial error and got an unfriendly error message:
~~~java
Test testCheckpointRescalingInKeyedState(org.apache.flink.test.checkpointing.RescaleCheckpointManuallyITCase) failed with:
java.util.NoSuchElementException: No value present
	at java.util.Optional.get(Optional.java:135)
	at org.apache.flink.test.checkpointing.RescaleCheckpointManuallyITCase.runJobAndGetCheckpoint(RescaleCheckpointManuallyITCase.java:166)
	at org.apache.flink.test.checkpointing.RescaleCheckpointManuallyITCase.testCheckpointRescalingKeyedState(RescaleCheckpointManuallyITCase.java:129)
	at org.apache.flink.test.checkpointing.RescaleCheckpointManuallyITCase.testCheckpointRescalingInKeyedState(RescaleCheckpointManuallyITCase.java:107)
~~~
I think we should still improve this.

Reply:
Thanks for pointing this out, it happens because checkpoint in `NotifyingDefiniteKeySource` completed does not mean the checkpoint of the whole job is completed, I updated the wait condition in `NotifyingDefiniteKeySource` to fix this.

Reply:
I did not come across the rare condition, I just use another fake exception to get the unfriendly error message. And it seems the error still exists after your changes.

Reply:
> I just use another fake exception to get the unfriendly error message.
Do you mean  the artificial error :
```Java
 if (failAfterEmission) {
         throw new FlinkRuntimeException(
           "Make job fail artificially, to retain completed checkpoint.");
```
Do you have any suggestions to improve this?

> And it seems the error still exists after your changes.
I run this test 10 times in my local env with unaligned checkpoint enabled, and didn't see "No value present" error. Do you know the root cause of this error?

Reply:
You can throw exceptions during the operator initialization. The error message of "No value present" was thrown due to the job did not ever run successfully.

Reply:
Oh, I see. I modify the error message when the latest completed checkpoint does not exist.